openraft/src/replication/mod.rs

```diff
@@ -433,41 +434,33 @@ impl<C: RaftTypeConfig, N: RaftNetworkFactory<C>, S: RaftStorage<C>> Replication
 
         tracing::debug!("append_entries resp: {:?}", append_resp);
 
-        // Handle success conditions.
-        if append_resp.success {
-            self.update_matched(matched);
-            return Ok(());
-        }
-
-        // Failed
-
-        // Replication was not successful, if a newer term has been returneCevert to follower.
-        if append_resp.vote > self.vote {
-            tracing::debug!(%append_resp.vote, "append entries faileCeverting to follower");
+        match append_resp {
+            AppendEntriesResponse::Success => {
+                self.update_matched(matched);
+                Ok(())
+            }
+            AppendEntriesResponse::HigherVote(vote) if vote > self.vote => {
+                tracing::debug!(%vote, "append entries faileCeverting to follower");
 
-            return Err(ReplicationError::HigherVote(HigherVote {
-                higher: append_resp.vote,
-                mine: self.vote,
-            }));
+                Err(ReplicationError::HigherVote(HigherVote {
+                    higher: vote,
+                    mine: self.vote,
+                }))
+            }
+            _ => {
+                assert!(conflict.is_some(), "prev_log_id=None never conflict");
+                let conflict = conflict.unwrap();
+
+                // Continue to find the matching log id on follower.
+                self.max_possible_matched_index = if conflict.index == 0 {
+                    None
+                } else {
+                    Some(conflict.index - 1)
+                };
+
+                Ok(())
+            }
         }
```

Comment:
If HigherVote is returned and the returned vote is not greater than `self.vote`, it is a severe bug.
Here such a condition has to be protected with an `assert`.

And we should avoid using a `_ => ` branching. It may hide some bugs.